### PR TITLE
[SPARK-17180] [SQL] Fix View Resolution Order in ALTER VIEW AS SELECT

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

In the current master branch, when users do not specify the database name in the `ALTER VIEW AS SELECT` command, we always try to alter the permanent view even if the temporary view exists. This PR is to resolve this issue. 

After the fix, `ALTER VIEW AS SELECT` command alters the temporary view if the temp view exists; otherwise, try to alter the permanent view. This order is consistent with another command `DROP VIEW`, since users are unable to specify the keyword TEMPORARY.
### How was this patch tested?

Added test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 1]
```diff
@@ -105,7 +105,13 @@ case class CreateViewCommand(
     }
     val sessionState = sparkSession.sessionState
 
-    if (isTemporary) {
+    // 1) CREATE VIEW: create a temp view when users explicitly specify the keyword TEMPORARY;
+    //                 otherwise, create a permanent view no matter whether the temporary view
+    //                 with the same name exists or not.
+    // 2) ALTER VIEW: alter the temporary view if the temp view exists; otherwise, try to alter
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: also mention that we will alter a temp view when users explicitly specify the keyword TEMPORARY

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 2]
```diff
@@ -204,10 +206,67 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+
+  test("ALTER VIEW: alter a persistent view when a temp VIEW with same name exists") {
+    verifyAlterViewWithIdenticalName(isTempAlteredView = false)
+  }
+
+  private def verifyAlterViewWithIdenticalName (isTempAlteredView: Boolean) = {
+    withView("testView", "default.testView") {
+      val catalog = spark.sessionState.catalog
+      val oldViewQuery = "SELECT id FROM jt"
+      val newViewQuery = "SELECT  id, id1 FROM jt"
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: nit: remove extra spaces

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 3]
```diff
@@ -204,10 +206,67 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
-  test("should allow CREATE permanent VIEW when a TEMPORARY VIEW with same name exists") {
+  test("ALTER VIEW: alter a temporary view when a permanent VIEW with same name exists") {
+    verifyAlterViewWithIdenticalName(isTempAlteredView = true)
+  }
+
+  test("ALTER VIEW: alter a persistent view when a temp VIEW with same name exists") {
+    verifyAlterViewWithIdenticalName(isTempAlteredView = false)
+  }
+
+  private def verifyAlterViewWithIdenticalName (isTempAlteredView: Boolean) = {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: how about `alterTempView`

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: Sure, will change it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 4]
```diff
@@ -204,10 +206,67 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+
+      val persistentView = catalog.getTableMetadata(
+        TableIdentifier(table = "testView", database = Some("default")))
+      assert(persistentView.tableType == CatalogTableType.VIEW)
+      val tempView = catalog.getTableMetadata(TableIdentifier("testView"))
+      assert(tempView.tableType == CatalogTableType.VIEW)
+      assert(tempView.viewOriginalText.isEmpty)
+
+      if (isTempAlteredView) {
+        // View Text of the persistent view default.testView is changed
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: `is changed` -> `is not changed`

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: : )

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: question: how can you tell whether it's CREATE VIEW or ALTER VIEW?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Yeah! The only way is to pass a flag.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: it already has 3 flags: `allowExisting`, `replace`, `isTemporary`. We should rethink about it and decide what flags we really need for this command.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: True, let me think about it. Thanks!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]:
```SQL
      CREATE (OR REPLACE)? TEMPORARY? VIEW (IF NOT EXISTS)? tableIdentifier
        identifierCommentList? (COMMENT STRING)?
        (PARTITIONED ON identifierList)?
        (TBLPROPERTIES tablePropertyList)? AS query
```

Each flag corresponds to a keyword in the `CREATE VIEW DDL` command. 

- `OR REPLACE` -> `replace`
- `TEMPORARY ` -> `isTemporary `
- `IF NOT EXISTS` -> `allowExisting`

If we want to use the same command `CreateViewCommand` to process both `CREATE VIEW` and `ALTER VIEW`, it sounds reasonable to add a new flag. So far, I have not found a way to combine them.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: like `CreateMode`, can we use `SaveMode` for `replace` and `allowExisting`?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: uh, I got it. : ) Copied what @liancheng wrote one month ago. 

allowExisting|replace|SaveMode
-------------|-------|--------
true|	false	|Ignore
false	|false|ErrorIfExists
false|true|Overwrite
true|true|Overwrite

Let me try it. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -1294,20 +1296,29 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
       userSpecifiedColumns: Seq[(String, Option[String])],
       query: QueryContext,
       properties: Map[String, String],
-      allowExisting: Boolean,
+      ignoreIfExists: Boolean,
       replace: Boolean,
-      isTemporary: Boolean): LogicalPlan = {
+      isTemporary: Boolean,
+      isAlterViewAsSelect: Boolean): LogicalPlan = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: how about a `viewType` parameter which is an enum?
```
public enum SaveMode {
  Temporary,
  Permanent,
  Either
}
```
We are going to add `global temp view` soon, so this enum will also be useful at that time.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Sure, will do it. Thanks!

File: sql/core/src/main/java/org/apache/spark/sql/ViewType.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql;
+
+/**
+ * ViewType is used to specify the type of views.
+ */
+public enum ViewType {
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/ViewType.java]: This doesn't need to be public to end users, we can put it in `view.scala` and use `sealed trait` to implement it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 7]
```diff
@@ -69,23 +66,17 @@ case class CreateViewCommand(
 
   override def output: Seq[Attribute] = Seq.empty[Attribute]
 
-  if (!isTemporary) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: why remove this check?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 8]
```diff
@@ -105,7 +96,14 @@ case class CreateViewCommand(
...
 
-    if (isTemporary) {
+    // 1) CREATE VIEW: create a temp view when users explicitly specify the keyword TEMPORARY;
+    //                 otherwise, create a permanent view no matter whether the temporary view
+    //                 with the same name exists or not.
+    // 2) ALTER VIEW: alter the temporary view if the temp view exists; otherwise, try to alter
+    //                the permanent view. Here, it follows the same resolution like DROP VIEW,
+    //                since users are unable to specify the keyword TEMPORARY.
+    if (viewType == ViewType.Temporary ||
+        (viewType != ViewType.Permanent && sessionState.catalog.isTemporaryTable(name))) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: This is not so readable, how about we use 3 branches for temp view and permanent view and any view individually?

Also, we don't need to mention CREATE VIEW or ALTER VIEW here, the semantic is clearly defined by the SaveMode and ViewType, we just need to document how these codes match the semantic.

File: sql/core/src/main/java/org/apache/spark/sql/ViewType.java (hunk 6 above)
Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/ViewType.java]: I thought you want me to use public enum. Let me change it now. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: `originalText` is empty only when the view is created through Dataset APIs, as shown in the [code comment ](https://github.com/gatorsmile/spark/blob/5606344584c1c45cf49d2a95158c55193e777a9e/sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala#L40-L41)

This checking looks useless. If you think we should keep it, I can add it back. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Sure, let me change it. Thanks!

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
after thinking about it more, I think it's better to use a different code path for ALTER VIEW, instead of sharing the same implementation of CREATE VIEW. Do you mind review https://github.com/apache/spark/pull/14874? thanks!

**Comment by gatorsmile** [comment]
Yeah, I am also fine about splitting them into two commands. Let me close it now.

**Pull request closed without merging**